### PR TITLE
Create file setting

### PR DESCRIPTION
**Pull request by rastytheamateur**
Add new `FileSetting` and corresponding `FileValidator`.

### CODE REVIEW

**Comment by pawamoy** [comment]
I like it :+1:

File: src/appsettings/validators.py [hunk 1]
```diff
@@ -67,3 +69,43 @@ def __call__(self, value):
...
+        self.exists = exists
+        self.mode = mode
+        self.not_exist_message = not_exist_message or self.not_exist_message
+        self.no_perms_message = no_perms_message or self.no_perms_message
+        self.unexpected_error_message = unexpected_error_message or self.unexpected_error_message
+
+    def __call__(self, value):
+        try:
+            with open(value, mode=self.mode):
+                pass
```
Comment (stinovlas) [inline src/appsettings/validators.py:103]: You could just call `open(value, mode=self.mode).close()`. There's no need tu use a context manager when you don't do anything.

File: src/appsettings/validators.py [hunk 2]
```diff
@@ -67,3 +69,43 @@ def __call__(self, value):
...
+        except IOError as e:
+            if e.errno == errno.ENOENT:
+                if self.exists:
+                    raise ValidationError(self.not_exist_message, params={"value": value})
+            elif e.errno == errno.EACCES:
+                raise ValidationError(self.no_perms_message, params={"value": value, "mode": self.mode})
+            else:
+                raise ValidationError(self.unexpected_error_message, params={"value": value, "exception": e})
```
Comment (stinovlas) [inline src/appsettings/validators.py:111]: I think it would be more elegant to catch individual errors such as `FileNotFoundError` rather than having catch-all and if.

File: tests/test_appsettings.py [hunk 3]
```diff
@@ -19,21 +22,30 @@ class SettingTestCase(SimpleTestCase):
     def _imported_object2():
         return "nothing"
 
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        fd, cls.temporary_file_name = mkstemp()
```
Comment (stinovlas) [inline tests/test_appsettings.py:28]: `mkstemp` opens the file as well. Should we not at least close it after the creation?

File: tests/test_appsettings.py [hunk 4]
```diff
@@ -19,21 +22,30 @@ class SettingTestCase(SimpleTestCase):
...
         self.message_missing_item = "%s setting is missing required item"
         self.message_no_attr = "has no attribute '%s'"
 
     def test_setting(self):
-        setting = appsettings.Setting(name='simple')
-        assert setting.name == 'simple'
-        assert setting.full_name == 'SIMPLE'
+        setting = appsettings.Setting(name="simple")
+        assert setting.name == "simple"
+        assert setting.full_name == "SIMPLE"
```
Comment (stinovlas) [inline tests/test_appsettings.py:42]: This seems to be an unrelated change that has nothing to do with `FileSetting`.

File: src/appsettings/validators.py [hunk 5]
```diff
@@ -67,3 +69,43 @@ def __call__(self, value):
...
+            mode (str): ``mode`` argument for ``open`` builtin function.
+            not_exist_message (str): Error message (containing ``value``)
+            no_perms_message (str): Error message (containing ``value`` and ``mode``)
+            unexpected_error_message (str): Error message (containing ``value`` and ``exception``)
+        """
+        self.exists = exists
+        self.mode = mode
+        self.not_exist_message = not_exist_message or self.not_exist_message
+        self.no_perms_message = no_perms_message or self.no_perms_message
+        self.unexpected_error_message = unexpected_error_message or self.unexpected_error_message
```
Comment (ziima) [inline src/appsettings/validators.py:98]: When there are multiple messages, I'd rather have them joined in a dictionary.

File: src/appsettings/settings.py [hunk 6]
```diff
@@ -967,3 +974,41 @@ def check(self):
...
+
+
+class FileSetting(Setting):
+    """
+    File setting.
+
+    Value of this setting is a pathlib.Path instance.
+    """
+
+    def __init__(self, exists=True, mode="r", *args, **kwargs):
```
Comment (ziima) [inline src/appsettings/settings.py:986]: Custom arguments are added to the end, not beginning.

File: src/appsettings/settings.py [hunk 7]
```diff
@@ -967,3 +974,41 @@ def check(self):
...
+        """
+        super(FileSetting, self).__init__(*args, **kwargs)
+        self.exists = exists
+        self.mode = mode
+        self.validators.append(FileValidator(exists, mode))
+
+    def transform(self, value):
+        """Transform value to Path instance."""
+        if not isinstance(value, Path):
+            value = Path(str(super(FileSetting, self).transform(value)))
```
Comment (ziima) [inline src/appsettings/settings.py:1013]: * Why is there the `str`?
* The parent should be called in all cases.

File: src/appsettings/validators.py [hunk 8]
```diff
@@ -67,3 +69,43 @@ def __call__(self, value):
...
+        Try to open a file using a given ``mode``.
+
+        Args:
+            exists (bool): Require existence of the file
+            mode (str): ``mode`` argument for ``open`` builtin function.
+            not_exist_message (str): Error message (containing ``value``)
+            no_perms_message (str): Error message (containing ``value`` and ``mode``)
+            unexpected_error_message (str): Error message (containing ``value`` and ``exception``)
+        """
+        self.exists = exists
```
Comment (ziima) [inline src/appsettings/validators.py:94]: If `exists` is False, then the validator doesn't make any sense. If I don't want to check the existence of a file, then I shouldn't use the validator at all.

File: tests/test_appsettings.py (hunk 3 above)
Comment (ziima) [reply tests/test_appsettings.py:28]: * I suggest using `NamedTemporaryFile` instead. It deletes the file after usage.
* To reduce a possibility of a test interference, I'd use different file for each test, i.e. move this into a `setUp`.

File: tests/test_appsettings.py (hunk 4 above)
Comment (ziima) [reply tests/test_appsettings.py:42]: There are more of those.

File: tests/test_appsettings.py [hunk 9]
```diff
@@ -425,6 +440,20 @@ def test_nested_dict_setting_required_both_inner_and_outer_setting(self):
...
+            setting.check()
+            assert isinstance(setting.value, Path)
+            assert setting.value.samefile(self.temporary_file_name)
+
+    def test_file_setting_path(self):
+        setting = appsettings.FileSetting(name="file")
+        with override_settings(FILE=Path(self.temporary_file_name)):
+            setting.check()
+            assert isinstance(setting.value, Path)
+            assert setting.value.samefile(self.temporary_file_name)
```
Comment (ziima) [inline tests/test_appsettings.py:455]: I'd create a separate test case for the `FileSetting`. It seems the `SettingTestCase` have grown too big already.

File: tests/test_validators.py [hunk 10]
```diff
@@ -51,3 +62,56 @@ def test_valid(self):
...
+    def tearDownClass(cls):
+        os.remove(cls.temporary_file_name)
+
+    def test_valid_read(self):
+        FileValidator()(self.temporary_file_name)
+
+    def test_valid_write(self):
+        FileValidator(mode="w")(self.temporary_file_name)
+
+    @patch("appsettings.validators.open", side_effect=IOError(errno.ENOENT, "Not a file"))
```
Comment (ziima) [inline tests/test_validators.py:83]: This could be easily mocked by creating an empty temporary directory.

File: tests/test_validators.py [hunk 11]
```diff
@@ -51,3 +62,56 @@ def test_valid(self):
...
+    def test_file_does_not_exist(self, open_mock):
+        with self.assertRaisesRegex(ValidationError, r"File .* does not exist\."):
+            FileValidator()(self.temporary_file_name)
+
+    @patch("appsettings.validators.open", side_effect=IOError(errno.ENOENT, "Not a file"))
+    def test_file_does_not_exist_without_validation(self, open_mock):
+        FileValidator(exists=False)(self.temporary_file_name)
+
+    @patch("appsettings.validators.open", side_effect=IOError(errno.EACCES, "No required permissions"))
```
Comment (ziima) [inline tests/test_validators.py:92]: We could actually modify the permissions of a test file.

File: src/appsettings/validators.py [hunk 12]
```diff
@@ -67,3 +69,43 @@ def __call__(self, value):
...
+        self.mode = mode
+        self.not_exist_message = not_exist_message or self.not_exist_message
+        self.no_perms_message = no_perms_message or self.no_perms_message
+        self.unexpected_error_message = unexpected_error_message or self.unexpected_error_message
+
+    def __call__(self, value):
+        try:
+            with open(value, mode=self.mode):
```
Comment (ziima) [inline src/appsettings/validators.py:102]: It's not possible to check if the path is a directory.

File: tests/test_appsettings.py (hunk 3 above)
Comment (stinovlas) [reply tests/test_appsettings.py:28]: I disagree with the `setUp`. Majority of tests doesn't need the temp file at all. If anything, we could transfer this operation to the tests themselves, or separate them into another `TestCase`.

File: src/appsettings/settings.py [hunk 13]
```diff
@@ -967,3 +974,41 @@ def check(self):
...
+            name (str): the name of the setting.
+            default (object): default value given to the setting.
+            required (bool): whether the setting is required or not.
+            prefix (str):
+                the setting's prefix (overrides ``AppSettings.Meta`` prefix).
+            call_default (bool): whether to call the default (if callable).
+            transform_default (bool): whether to transform the default value.
+            validators (list of callables): list of additional validators to use.
+        """
+        super(FileSetting, self).__init__(*args, **kwargs)
```
Comment (ziima) [inline src/appsettings/settings.py:1005]: Since we use python 3 only, please use only `super()` and add annotations to the new code.

File: tests/test_appsettings.py (hunk 4 above)
Comment (rastytheamateur) [reply tests/test_appsettings.py:42]: Those are changes from `black` and `flake8` because of `tox`. Should I remove the changes and let the `tox` fail?

File: src/appsettings/settings.py (hunk 7 above)
Comment (rastytheamateur) [reply src/appsettings/settings.py:1013]: > * Why is there the `str`?

We could get a `(Pure)PosixPath`, for example. But I have no problem with removing it.

File: src/appsettings/validators.py (hunk 8 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py:94]: I disagree. You may want to open a non-existing file with write mode in a folder where you have no write permissions. The open then fails with `PermissionError`.

`open('no-w-permissions-dir/example.txt', mode='r')`

File: src/appsettings/validators.py (hunk 12 above)
Comment (rastytheamateur) [inline src/appsettings/validators.py:102]: One problem which I see is that it really opens and closes the file, so if there was a file with a content, this check removes the content no matter if you really use the file with write mode in your script. In this way the previous validation was better, safer (I mean the one with `os.access()`)

Comment (rastytheamateur) [reply src/appsettings/validators.py:102]: Well, I don't think it should be a directory... For a directory we could create a new `Setting`.

If you do want to make both files and directories setting, I would suggest to create the previous validation based on `os.access()`. Another reason to do it might be [my comment below](https://github.com/pawamoy/django-appsettings/pull/81/#pullrequestreview-386395778). But still - I would rather have two individual subsetting classes.

File: tests/test_validators.py (hunk 11 above)
Comment (rastytheamateur) [reply tests/test_validators.py:92]: I am not sure how would you do it. I haven't found a way to mock permissions. Changing the file itself may result in impossibility to remove it, or not?

File: src/appsettings/validators.py (hunk 5 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py:98]: You mean to pack all the variables into one dict?
```
class FileValidator(object):

    error_messages = dict(
        not_exist_message = "...",
        no_perms_message = "...",
        unexpected_error_message = "...",
    )

    def __init__(..., error_messages=None):
        self.error_messages = error_messages or self.error_messages
```
But then you are not able to modify just one error message.

File: tests/test_appsettings.py (hunk 4 above)
Comment (ziima) [reply tests/test_appsettings.py:42]: It doesn't seem to be failing on master.

File: src/appsettings/validators.py (hunk 5 above)
Comment (ziima) [reply src/appsettings/validators.py:98]: > You mean to pack all the variables into one dict?

Yes.

> But then you are not able to modify just one error message.

You can merge the dictionaries.

File: src/appsettings/settings.py (hunk 7 above)
Comment (ziima) [reply src/appsettings/settings.py:1013]: Then the condition above would fail, wouldn't it?

File: tests/test_appsettings.py (hunk 9 above)
Comment (rastytheamateur) [reply tests/test_appsettings.py:455]: There are only two tests for `FileSetting`, it is probably the last setting to separate. I would separate all the tests for each setting type in another ticket/branch.

File: src/appsettings/validators.py (hunk 8 above)
Comment (ziima) [reply src/appsettings/validators.py:94]: It doesn't seem that `exists` would have any effect on that case. It will fail with `no_perms_message` regardless of `exists` value flag.

File: tests/test_validators.py (hunk 11 above)
Comment (ziima) [reply tests/test_validators.py:92]: I don't think so. To delete a file, you need a write permission on its directory.

File: tests/test_appsettings.py (hunk 3 above)
Comment (ziima) [reply tests/test_appsettings.py:28]: +1 for separating the tests.

File: src/appsettings/validators.py (hunk 5 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py:98]: > You can merge the dictionaries.

I don't like it. In your case there will have to be some dictionary keys in the documentation and all becomes less clean. Should I throw an exception on invalid key? Should I ignore it?

File: tests/test_appsettings.py (hunk 9 above)
Comment (ziima) [reply tests/test_appsettings.py:455]: We should do that as well. But addition of a new tests is great place to start with separation.

File: src/appsettings/settings.py (hunk 7 above)
Comment (rastytheamateur) [reply src/appsettings/settings.py:1013]: True, forgot about the condition.

File: src/appsettings/validators.py (hunk 12 above)
Comment (ziima) [reply src/appsettings/validators.py:102]: It seems they would be rather duplicated. I don't really see what would be the difference between setting for a file and a directory.

File: src/appsettings/validators.py (hunk 8 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py:94]: > If I don't want to check the existence of a file, then I shouldn't use the validator at all.

But you want to use the validator despite you don't care about the existence.

File: src/appsettings/validators.py (hunk 12 above)
Comment (ziima) [reply src/appsettings/validators.py:102]: Good point. That really doesn't look like something you want your settings should be doing. I would definitely avoid that.

File: src/appsettings/validators.py (hunk 5 above)
Comment (ziima) [reply src/appsettings/validators.py:96]: Ignore. It's a common approach in Django for error messages, so it's not something entirely unexpected.

File: src/appsettings/validators.py (hunk 8 above)
Comment (ziima) [reply src/appsettings/validators.py:92]: Exactly. I don't see the case, where the `exists` would make a difference, if I can just not use the validator at all.

File: tests/test_appsettings.py (hunk 4 above)
Comment (pawamoy) [reply tests/test_appsettings.py]: It's because CI on GitLab was temporarily disabled.

File: src/appsettings/validators.py (hunk 1 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: I decided to use the `os.access` validation. `open(file, mode="w")` erased the file content.

File: src/appsettings/validators.py (hunk 12 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: Should be all right now.

File: src/appsettings/settings.py (hunk 13 above)
Comment (rastytheamateur) [reply src/appsettings/settings.py]: It should be done now.

File: src/appsettings/validators.py (hunk 12 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: When I see `FileSetting`, I expect it to be a file, not a directory. When I see `DirectorySetting`, I expect it to be a directory.

But OK, not it should be able to take a dir as well.

File: tests/test_validators.py (hunk 11 above)
Comment (rastytheamateur) [reply tests/test_validators.py]: You are right, done. But I am not sure if is it the right way. Anything can go wrong and some ghost files and directories may remain in the file system (experienced during development).

File: tests/test_validators.py (hunk 10 above)
Comment (rastytheamateur) [reply tests/test_validators.py]: Done.

File: tests/test_appsettings.py (hunk 9 above)
Comment (rastytheamateur) [reply tests/test_appsettings.py]: Okey, I will do it, but it seems to be so strange to me. At this point there is one huge class `SettingTestCase` and one poor little class `FileSettingTestCase` with two tests.

Much better and cleaner would be to do it all at once, by my opinion.

File: src/appsettings/settings.py (hunk 6 above)
Comment (rastytheamateur) [reply src/appsettings/settings.py]: Done.

Do I really have to name all the parameters of the parent class, or is there a nicer way to do it?

File: src/appsettings/validators.py (hunk 5 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: Done.

File: tests/test_appsettings.py (hunk 3 above)
Comment (rastytheamateur) [reply tests/test_appsettings.py]: Operation transferred to tests themselves, tests separated to a special VIP class.

File: src/appsettings/validators.py (hunk 2 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: I decided to use the `os.access` validation. `open(file, mode="w")` erased the file content.

File: docs/api.rst [hunk 14]
```diff
@@ -28,6 +28,9 @@ API reference
 .. autoclass:: appsettings.PositiveFloatSetting
     :members:
 
+.. autoclass:: appsettings.FileSetting
+    :members:
+
```
[the remark this answers is not in the document]
Comment (rastytheamateur) [reply docs/api.rst]: Manually. Moved between `PositiveIntegerSetting` and `FloatSetting`.

File: src/appsettings/settings.py [hunk 15]
```diff
@@ -967,3 +974,41 @@ def check(self):
...
+
+    def __init__(self, exists=True, mode="r", *args, **kwargs):
+        """
+        Initialization method.
+
+        Note that attributes ``exists`` and ``mode`` are used in ``FileValidator``. For more information check its
+        documentation.
+
+        Args:
+            exists (bool): File already exists.
```
[the remark this answers is not in the document]
Comment (rastytheamateur) [reply src/appsettings/settings.py]: Docstring changed to `Check whether or not does the file exist.`

File: tests/test_appsettings.py (hunk 4 above)
Comment (rastytheamateur) [reply tests/test_appsettings.py]: I was a bit behind `pawamoy/master`, pulled and rebased. But there were still 3 changes in tests, which had to be fixed.

Comment (ziima) [reply tests/test_appsettings.py]: I'd leave that for some cleaning commit/PR.

File: src/appsettings/settings.py (hunk 6 above)
Comment (ziima) [reply src/appsettings/settings.py]: Nope.

File: src/appsettings/settings.py [hunk 16]
```diff
@@ -1153,3 +1160,62 @@ def check(self):
...
+            validators=validators,
+        )
+        self.exists = exists
+        self.validators.append(FileValidator(exists, read_perm, write_perm))
+
+    def transform(self, value):
+        """Transform value to Path instance."""
+        value = super().transform(value)
+        if not isinstance(value, Path):
+            value = Path(value)
```
Comment (ziima) [inline src/appsettings/settings.py:1220]: It seems the `Path` can wrap itself. so the condition is not necessary.

File: tests/test_validators.py (hunk 11 above)
Comment (ziima) [reply tests/test_validators.py]: That seems rather unlikely.

File: src/appsettings/validators.py (hunk 12 above)
Comment (ziima) [reply src/appsettings/validators.py]: Well, technically directory is also a file :)

File: src/appsettings/validators.py [hunk 17]
```diff
@@ -67,3 +69,53 @@ def __call__(self, value):
...
+        return directory
+
+    def __call__(self, value):
+        """Validate the ``value``."""
+        if not os.path.exists(value):
+            if self.exists:
+                raise ValidationError(self.error_messages["exists"], params={"value": value})
+            # If file does not exist (yet), we might want to check at least permissions to the directory, where the file
+            # will be located.
+            value = self._get_nearest_existing_directory(value)
```
Comment (ziima) [inline src/appsettings/validators.py:115]: This seems to be rather far-fetched. We have no idea what are the requirements for the directories up in hierarchy. I'd rather skip the check entirely and leave it on a custom validator if required.

File: src/appsettings/validators.py [hunk 18]
```diff
@@ -67,3 +69,53 @@ def __call__(self, value):
...
+        if self.read_perm:
+            if not os.access(value, os.R_OK):
+                raise ValidationError(self.error_messages["read_perm"], params={"value": value})
+        if self.write_perm:
+            if not os.access(value, os.W_OK):
+                raise ValidationError(self.error_messages["write_perm"], params={"value": value})
```
Comment (ziima) [inline src/appsettings/validators.py:121]: Executable permission is missing.

File: src/appsettings/validators.py (hunk 17 above)
Comment (ziima) [reply src/appsettings/validators.py:115]: Also, once we start looking at the directories up, we need to check completely different set of permissions.

Comment (rastytheamateur) [reply src/appsettings/validators.py:115]: I don't understand why is it so bad and which different permissions should we check? When you want to read/write a new file in a directory, you just need read/write permission in the directory, or not?

So what do you suggest? Skip the `read_perm` and `write_perm` when the file does not exist without any message to the user?

Comment (ziima) [reply src/appsettings/validators.py:115]: If I need to read a file that may not exist, I need to check the write permission on its directory if I want to create it.

Comment (rastytheamateur) [reply src/appsettings/validators.py:115]: But you don't want to create it, when you want to just read it. When you want to have read permission on an non-existing file, you just expect that a file with read permission will have been created when you try to read it. Read permission on an existing parent directory does not ensure it, but it is a good sign.

File: tests/test_appsettings.py (hunk 4 above)
Comment (rastytheamateur) [reply tests/test_appsettings.py]: Another pull seems to have fixed it.

File: src/appsettings/settings.py (hunk 16 above)
Comment (rastytheamateur) [reply src/appsettings/settings.py]: Fixed.

File: src/appsettings/validators.py (hunk 17 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: Fixed according to our consultation.

File: src/appsettings/validators.py (hunk 18 above)
Comment (rastytheamateur) [reply src/appsettings/validators.py]: Fixed.

File: src/appsettings/validators.py [hunk 19]
```diff
@@ -67,3 +69,29 @@ def __call__(self, value):
             if not isinstance(element, self.value_type):
                 params = {"key": key, "value": element, "type": self.value_type.__name__}
                 raise ValidationError(self.message, params=params)
+
+
+class FileValidator(object):
+    """Validator which checks file existence and permissions."""
+
+    DEFAULT_ERROR_MESSAGE = "No required permissions for the file %(value)s."
```
[the remark this answers is not in the document]
Comment (ziima) [reply src/appsettings/validators.py:77]: I'd use `Insufficient permissions ...` or rephrase the message to `The file %(value)s doesn't have required permissions.` or similar.

File: src/appsettings/settings.py [hunk 20]
```diff
@@ -1153,3 +1160,56 @@ def check(self):
...
+        super().__init__(
+            name=name,
+            default=default,
+            required=required,
+            prefix=prefix,
+            call_default=call_default,
+            transform_default=transform_default,
+            validators=validators,
+        )
+        self.validators.append(FileValidator(mode))
```
[the remark this answers is not in the document]
Comment (rastytheamateur) [reply src/appsettings/settings.py]: Fixed.

File: src/appsettings/validators.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (rastytheamateur) [reply src/appsettings/validators.py]: Fixed.

File: src/appsettings/validators.py [hunk 21]
```diff
@@ -67,3 +69,29 @@ def __call__(self, value):
...
+    def __init__(self, mode, error_message=None):
+        """
+        Validator initialization.
+
+        Args:
+            mode (optional int): Required permission for the file. None means no validation (f.e. file does not exist
+                yet), other values might consist of inclusive OR between F_OK, R_OK, W_OK and X_OK constants. When
+                checking read, write or execute permission, file existence (F_OK) is implicitly required. For more
+                information check ``os.access`` documentation (https://docs.python.org/3/library/os.html#os.access).
+            error_message (str): Override default ``ValidationError`` message. The message must contain ``%(value)s``.
```
[the remark this answers is not in the document]
Comment (rastytheamateur) [reply src/appsettings/validators.py]: I thought `ValidationError` wouldn't be satisfied, because `params` attribute would contain invalid values. I probably confused it with another error or string formatter.

Fixed.

File: src/appsettings/validators.py [hunk 22]
```diff
@@ -67,3 +69,29 @@ def __call__(self, value):
...
+
+    def __init__(self, mode, error_message=None):
+        """
+        Validator initialization.
+
+        Args:
+            mode (optional int): Required permission for the file. None means no validation (f.e. file does not exist
+                yet), other values might consist of inclusive OR between F_OK, R_OK, W_OK and X_OK constants. When
+                checking read, write or execute permission, file existence (F_OK) is implicitly required. For more
+                information check ``os.access`` documentation (https://docs.python.org/3/library/os.html#os.access).
```
[the remark this answers is not in the document]
Comment (rastytheamateur) [reply src/appsettings/validators.py]: Fixed.

File: src/appsettings/validators.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ziima) [reply src/appsettings/validators.py]: `ValidationError`s don't care :)

**Comment by pawamoy** [comment]
Thank you all for your contributions :tada:
